README.md
![FTPopOverMenu](https://github.com/liufengting/FTResourceRepo/blob/master/Resource/FTPopOverMenu_logo/logotype_1280.png?raw=true)

# FTPopOverMenu

[![Twitter](https://img.shields.io/badge/twitter-@liufengting-blue.svg?style=flat)](http://twitter.com/liufengting) 
[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/liufengting/FTPopOverMenu/master/LICENSE)
[![Version](https://img.shields.io/cocoapods/v/FTPopOverMenu.svg?style=flat)](http://cocoapods.org/pods/FTPopOverMenu)
[![Codebeat badge](https://codebeat.co/badges/763991fb-8307-4683-8b19-6c8aba9f1683)](https://codebeat.co/projects/github-com-liufengting-ftpopovermenu)
[![Download](https://img.shields.io/cocoapods/dt/FTPopOverMenu.svg?maxAge=2592000)](http://cocoapods.org/pods/FTPopOverMenu)
[![CocoaPods](https://img.shields.io/cocoapods/at/FTPopOverMenu.svg?maxAge=2592000)](http://cocoapods.org/pods/FTPopOverMenu)
[![CocoaPods](https://img.shields.io/cocoapods/metrics/doc-percent/FTPopOverMenu.svg?maxAge=2592000)](http://cocoapods.org/pods/FTPopOverMenu)
[![CI Status](http://img.shields.io/travis/liufengting/FTPopOverMenu.svg?style=flat)](https://travis-ci.org/liufengting/FTPopOverMenu)
[![GitHub stars](https://img.shields.io/github/stars/liufengting/FTPopOverMenu.svg)](https://github.com/liufengting/FTPopOverMenu/stargazers)
[![Gitter](https://badges.gitter.im/liufengting/FTPopOverMenu.svg)](https://gitter.im/liufengting/FTPopOverMenu?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)

`FTPopOverMenu` is a pop over menu for `iOS` which is maybe the easiest one to use, supports both `portrait` and `landscape`. It can show from any `UIView`, any `UIBarButtonItem` and any `CGRect`. Simplest APIs, enable you to change the style in one line of code.


### menu array supports following context:

* image name (NSString, only main bundle),
* image (UIImage),
* image remote URL string (NSString),
* image remote URL (NSURL),
* model (FTPopOverMenuModel, select state support)

### ⚠️ Big API Change from `1.3.6` to `2.0.0` (See [CHANGELOG](https://github.com/liufengting/FTPopOverMenu/blob/master/CHANGELOG.md) for more info)

* supports separated configuration. 
* change 'tintColor' to 'backgroundColor'

# About Swift Version

`FTPopOverMenu_Swift` is now published. [Take a look at it now](https://github.com/liufengting/FTPopOverMenu_Swift). The swift version has same APIs with this OC version.

# ScreenShots

![screenshots](https://raw.githubusercontent.com/liufengting/FTResourceRepo/master/Resource/FTPopOverMenu/screenshots.gif)

## Recently UrbanHooker added round arrow support, I changed a little bit.

![screenshots](https://raw.githubusercontent.com/liufengting/FTResourceRepo/master/Resource/FTPopOverMenu/FTPopOverMenuRoundArrow.png)

# Installation

## Manually

* clone this repo.
* Simply drop the '/FTPopOverMenu' folder into your project.
* import 'FTPopOverMenu.h'

## CocoaPods
`FTPopOverMenu` is available through [CocoaPods](http://cocoapods.org). To install it, simply add the following line to your Podfile:

```ruby
pod 'FTPopOverMenu'
```

# Usage

## Customize

Do any of the following settings to set the style (Only set the ones you want to change) before showing.

```objective-c
    FTPopOverMenuConfiguration *configuration = [FTPopOverMenuConfiguration defaultConfiguration];
    configuration.menuRowHeight = ...
    configuration.menuWidth = ...
    configuration.textColor = ...
    configuration.textFont = ...
    configuration.tintColor = ...
    configuration.borderColor = ...
    configuration.borderWidth = ...
    configuration.textAlignment = ...
    configuration.ignoreImageOriginalColor = ...;// set 'ignoreImageOriginalColor' to YES, images color will be same as textColor
    configuration.allowRoundedArrow = ...// Default is 'NO', if sets to 'YES', the arrow will be drawn with round corner.
    configuration.separatorColor = ...
    configuration.shadowColor = ... // Default is black
    configuration.shadowOpacity = ... // Default is 0 - choose anything between 0 to 1 to show actual shadow, e.g. 0.2
    configuration.shadowRadius = ... // Default is 5
    configuration.shadowOffsetX = ...
    configuration.shadowOffsetY = ...
```


## ~~setTintColor, default is gray color.~~ deprecated after 1.2.0

```objective-c
    [FTPopOverMenu setTintColor:[UIColor redColor]];
```
## ~~setTextColor, default is white color.~~ deprecated after 1.2.0

```objective-c
    [FTPopOverMenu setTextColor:[UIColor blackColor]];
```

## ~~setPreferedWidth, default is 120, should not be less than 50.~~ deprecated after 1.2.0

```objective-c
    [FTPopOverMenu setPreferedWidth:200];
```

## From SenderView, Menu Without Images.
 
```objective-c
    [FTPopOverMenu showForSender:sender
                        withMenu:@[@"MenuOne",@"MenuTwo",@"MenuThr"]
                       doneBlock:^(NSInteger selectedIndex) {
                           
                       } dismissBlock:^{
                          
                       }];
```

## From SenderView, Menu With Images.
 
```objective-c
    [FTPopOverMenu showForSender:sender
                        withMenu:@[@"MenuOne",@"MenuTwo",@"MenuThr"]
                  imageNameArray:@[@"setting_icon",@"setting_icon",@"setting_icon"]
                       doneBlock:^(NSInteger selectedIndex) {
                           
                       } dismissBlock:^{
                          
                       }];
```

## From SenderView, Menu With Images, custom configuration.
 
```objective-c
    [FTPopOverMenu showForSender:sender
                        withMenu:@[@"MenuOne",@"MenuTwo",@"MenuThr"]
                  imageNameArray:@[@"setting_icon",@"setting_icon",@"setting_icon"]
                   configuration:customConfiguration
                       doneBlock:^(NSInteger selectedIndex) {
                           
                       } dismissBlock:^{
                          
                       }];
```

## From SenderFrame/NavigationItem, Menu Without Images.
 
```objective-c
    [FTPopOverMenu showFromSenderFrame:CGRectMake(self.view.frame.size.width - 40, 20, 40, 40)
                              withMenu:@[@"123",@"234",@"345"]
                             doneBlock:^(NSInteger selectedIndex) {
                                 
                             } dismissBlock:^{
                                 
                             }];
```

## From SenderFrame/NavigationItem, Menu With Images.
 
```objective-c
    [FTPopOverMenu showFromSenderFrame:CGRectMake(self.view.frame.size.width - 40, 20, 40, 40)
                              withMenu:@[@"123",@"234",@"345"]
                        imageNameArray:@[@"setting_icon",@"setting_icon",@"setting_icon"]
                             doneBlock:^(NSInteger selectedIndex) {
                                 
                             } dismissBlock:^{
                                 
                             }];
```

## From SenderFrame/NavigationItem, Menu With Images, custom configuration.
 
```objective-c
    [FTPopOverMenu showFromSenderFrame:CGRectMake(self.view.frame.size.width - 40, 20, 40, 40)
                              withMenu:@[@"123",@"234",@"345"]
                        imageNameArray:@[@"setting_icon",@"setting_icon",@"setting_icon"]
	                     configuration:customConfiguration
                             doneBlock:^(NSInteger selectedIndex) {
                                 
                             } dismissBlock:^{
                                 
                             }];
```

## From barButtonItems .

- First: add action with event to you barButtonItems 

```objective-c
    [self.navigationItem setRightBarButtonItem:[[UIBarButtonItem alloc] initWithTitle:@"Edit" style:UIBarButtonItemStyleDone target:self action:@selector(onNavButtonTapped:event:)]];
```

- Second: implement the action, also supports custom configuration.

```objective-c
-(void)onNavButtonTapped:(UIBarButtonItem *)sender event:(UIEvent *)event
{

    [FTPopOverMenu showFromEvent:event
                        withMenu:@[@"123",@"234",@"345"]
                  imageNameArray:@[@"setting_icon",@"setting_icon",@"setting_icon"]
                  configuration:customConfiguration
                       doneBlock:^(NSInteger selectedIndex) {
                           
                       } dismissBlock:^{
                           
                       }];
}
```

# CHANGELOG

[CHANGELOG](https://github.com/liufengting/FTPopOverMenu/blob/master/CHANGELOG.md)

# More

* Logo Designed by @Tobaloidee
* Images in the demo from [https://www.iconfinder.com/iconsets/pokemon-go](https://www.iconfinder.com/iconsets/pokemon-go).
* If you like it, or maybe wanna use it in your own projects, please give me a `star`.
* Looking for another way of doing this? Take a look at my another [FTPopMenu at here](https://github.com/liufengting/FTPopMenu).  It was another way of showing pop over menus, which is wrapped from `UIPopoverViewController`, and more genetic and more good-looking. Feel free to try it.
 

<img src="https://raw.githubusercontent.com/liufengting/FTPopMenu/master/ScreenShots/ScreenShots2.png" width="300"/>


# License

`FTPopOverMenu` is available under the `MIT` license. See the `LICENSE` file for more info.




CHANGELOG.md
# FTPopOverMenu


| Version | Date | ChangeLog |
|:--------:|:--------:|:--------|
|1.0.0|2016-04-03| upload to CocoaPods|
|1.1.5|2016-08-04| fix a bug (arrowPoint.x)|
|1.1.6|2016-08-15| fix an issue of out of screen when too many menus|
|1.1.7|2016-08-22| fix an issue thanks to Cui Lizong|
|1.1.8|2016-08-25| add custom width support|
|1.1.9|2016-08-26| add custom text color support|
|1.2.0|2016-09-07| fix some bugs|
|1.2.1|2016-09-15| Big Change: more customizable|
|1.2.2|2016-10-12| fix a bug of customize height|
|1.2.3|2016-11-17| remove last cell's separator, now looks more clean|
|1.2.5|2016-11-24| add some more animations|
|1.2.6|2016-12-07| Menu text and icon margins can be configured via FTPopOverMenuConfiguration.|
|1.3.0|2016-12-21| Big jump. More image resource support |
|1.3.1|2016-12-21| fix bug: image cache name encode |
|1.3.2|2017-01-07| added round arrow support |
|1.3.3|2017-03-05| remove some unused methods |
|1.3.4|2018-01-26| add selected text color and cell background color |
|1.3.5|2018-06-15| Added separatorColor support. Thanks to @flyandi |
|1.3.6|2018-06-15| Added shadow support. Thanks to @flyandi |
|  | |  |
|2.0.0|2018-07-31| supports separated configuration. |
|2.0.1|2018-10-25| add cover background color. |
|2.0.2|2018-11-30| add menuCornerRadius and separatorInset. |
|2.0.3|2018-12-06| @Deeer: 修复异步回调时图片无法显示的问题 |
|2.0.4|2018-12-06| @Carles Estevadeordal: Added custom UIWindow support. |
|2.0.5|2019-08-02| @yooomark: Fix bug: set animationDuration in config do not work |
|2.0.6|2019-08-12| add customize icon size support |
|2.0.7|2019-08-30| correct selectedBackgroundColor, add cover color animation |
|2.0.8|2019-09-02| fix selectedBackgroundColor thanks to @Jean1992 |
|2.0.9|2019-10-23| add horizontalMargin |
|2.1.0|2020-02-21| add s.ios.deployment_target |
|2.1.1|2020-03-24| Fix: height wrong when too long. Fix: SafeArea height issue. |
|2.1.3|2022-12-08| Fix: dismiss animation interrupted occasionally |
|2.1.4|2022-12-08| update using projectf for files |




Demo/Demo/ViewController (1).m
//
//  ViewController.m
//  FTPopOverMenu
//
//  Created by liufengting on 16/4/5.
//  Copyright © 2016年 liufengting ( https://github.com/liufengting ). All rights reserved.
//

#import "ViewController.h"
#import "FTPopOverMenu.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];


    

}

- (IBAction)showMenuFromButton:(UIButton *)sender
{
    
    // Do any of the following setting to set the style (Only set what you want to change)

    
    // uncomment the following line to use custom settings
    
//#define USE_CUSTOM_SETTINGS

#ifdef USE_CUSTOM_SETTINGS
    FTPopOverMenuConfiguration *configuration = [FTPopOverMenuConfiguration defaultConfiguration];
    configuration.menuRowHeight = 40;
    configuration.menuWidth = 120;
    configuration.textColor = [UIColor orangeColor];
    configuration.textFont = [UIFont boldSystemFontOfSize:14];
    configuration.tintColor = [UIColor whiteColor];
    configuration.borderColor = [UIColor blackColor];
    configuration.borderWidth = 0.5;
//    configuration.textAlignment = NSTextAlignmentCenter;
//    configuration.ignoreImageOriginalColor = YES;// set 'ignoreImageOriginalColor' to YES, images color will be same as textColor

#endif
    
    
    [FTPopOverMenu showForSender:sender
                        withMenu:@[@"MenuOne",@"MenuTwo",@"MenuThree",@"MenuFour"]
                  imageNameArray:@[@"Pokemon_Go_01",@"Pokemon_Go_02",@"Pokemon_Go_03",@"Pokemon_Go_04"]
                       doneBlock:^(NSInteger selectedIndex) {
                           
                           NSLog(@"done block. do something. selectedIndex : %ld", (long)selectedIndex);
                           
                       } dismissBlock:^{
                           
                           NSLog(@"user canceled. do nothing.");
                           
                       }];
    
}



@end
